# Investigate Wine Quality DataSet
---

## Table of Contents
<ul>
    <li><a href="#intro">Introduction</a></li>  
    <li><a href="#wrangling">Data Wrangling</a></li>
</ul>    

<a id='intro'></a>
## Introduction
---
We're going to investigate this [dataset](https://archive.ics.uci.edu/ml/datasets/Wine+Quality) on physicochemical properties and quality ratings of red and white wine samples.

##  Wine Quality Data Set from UCI Machine Learning Lab
There are two datasets that provide information on samples of red and white variants of the Portuguese "Vinho Verde" wine.
Each sample of wine was rated for quality by wine experts and examined with physicochemical tests. Due to privacy and logistic
issues, only data on these physicochemical properties and quality ratings are available (e.g. there is no data about grape types,   wine brand, wine selling price, etc.). ([Source](https://archive.ics.uci.edu/ml/datasets/Wine+Quality))

### Attributes in Each Dataset:

**# Physicochemical Properties**
1. Fixed Acidity  
2. Volatile Acidity  
3. Citric Acid  
4. Residual Sugar  
5. Chlorides  
6. Free Sulfur Dioxide  
7. Total Sulfur Dioxide  
8. Density  
9. pH  
10. Sulphates  
11. Alcohol  

**# Quality Rating**  
12.  Quality - Score between 0 and 10 (median of at least 3 evaluations made by wine experts)

## Asking Questions

**Relevent questions to this dataset:**
 * What chemical characteristics are most important in predicting the quality of wine?
 * Is a certain type of wine (red or white) associated with higher quality?
 * Do wines with higher alcoholic content receive better ratings?
 * Do sweeter wines (more residual sugar) receive better ratings?
 * What level of acidity is associated with the highest quality?

In [1]:
# importing necessary packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
# Including a 'magic word' so that our visualizations are plotted inline with the notebook.
%matplotlib inline    
import seaborn as sns

<a id='wrangling'></a>
## Data Wrangling
---

### Gathering Data
Load wine quality datasets present in files `winequality-red.csv` and `winequality-white.csv`.

In [2]:
# loading data files
red_df = pd.read_csv('winequality-red.csv')
white_df = pd.read_csv('winequality-white.csv')
# printing first few rows
red_df.head()

fixed_acidity;volatile_acidity;citric_acid;residual_sugar;chlorides;free_sulfur_dioxide;total_sulfur-dioxide;density;pH;sulphates;alcohol;quality
0   7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5                                                                                               
1   7.8;0.88;0;2.6;0.098;25;67;0.9968;3.2;0.68;9.8;5                                                                                               
2  7.8;0.76;0.04;2.3;0.092;15;54;0.997;3.26;0.65;...                                                                                               
3  11.2;0.28;0.56;1.9;0.075;17;60;0.998;3.16;0.58...                                                                                               
4   7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5

Looks like columns in datasets are seprated by semicolon. Use  `sep = ';'`  parameter in  `read_csv`  method.

In [3]:
# load data files again with `sep` parameter 
red_df = pd.read_csv('winequality-red.csv', sep = ';')
white_df = pd.read_csv('winequality-white.csv', sep = ';')

### Assessing Data

In [4]:
print(red_df.shape)
red_df.head()

(1599, 12)


fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free_sulfur_dioxide  total_sulfur-dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

*In red wine dataframe column name `total_sulfur-dioxide` has `-`(hyphen) instead of `_`(underscore), which is different from others.*

In [5]:
print(white_df.shape)
white_df.head()

(4898, 12)


fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

*white-wine dataset has more samples comparing to red-wine dataset.*

lets check for any incorrect datatypes

In [6]:
# datatypes of features
white_df.dtypes

fixed_acidity           float64
volatile_acidity        float64
citric_acid             float64
residual_sugar          float64
chlorides               float64
free_sulfur_dioxide     float64
total_sulfur_dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object

In [7]:
red_df.dtypes

fixed_acidity           float64
volatile_acidity        float64
citric_acid             float64
residual_sugar          float64
chlorides               float64
free_sulfur_dioxide     float64
total_sulfur-dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object

*No incorrect datatypes*

lets check features with missing values

In [8]:
#features with missing values
red_df.isnull().sum()

fixed_acidity           0
volatile_acidity        0
citric_acid             0
residual_sugar          0
chlorides               0
free_sulfur_dioxide     0
total_sulfur-dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

In [9]:
white_df.isnull().sum()

fixed_acidity           0
volatile_acidity        0
citric_acid             0
residual_sugar          0
chlorides               0
free_sulfur_dioxide     0
total_sulfur_dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

*There are no features with missing values in both datasets.*

lets see if there are any duplicate rows in datasets

In [10]:
#duplicate rows in the dataset
white_df.duplicated().sum()

937

In [11]:
#duplicate rows in the red wine dataset
red_df.duplicated().sum()

240

*There are 937 and 240 duplicate rows in white wine and red wine datasets respectively.*